File: android/app/src/main/java/com/ilovetrains/app/OfflineRealtime.kt
package com.ilovetrains.app

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.coroutineScope
import org.json.JSONObject
import java.io.ByteArrayOutputStream
import java.io.InputStream
import java.net.HttpURLConnection
import java.net.URL
import java.time.Instant
import java.util.concurrent.ConcurrentHashMap

internal data class RealtimeResult<T>(val value: T, val observedAt: Long? = null, val matched: Boolean = false)
internal data class StopAssignment(val platform: String?, val stationId: String)

internal class OfflineRealtime {
    private data class StopUpdate(
        val stopId: String?,
        val sequence: Int?,
        val assignedStopId: String?,
        val arrival: Long?,
        val departure: Long?,
        val arrivalDelaySeconds: Int?,
        val departureDelaySeconds: Int?,
        val relationship: String,
    ) {
        fun matches(id: String, stopSequence: Int?): Boolean = when {
            stopId != null && sequence != null -> stopId == id && sequence == stopSequence
            stopId != null -> stopId == id
            sequence != null -> sequence == stopSequence
            else -> false
        }
    }

    private data class TripUpdate(
        val tripId: String,
        val serviceDate: String,
        val status: String,
        val delaySeconds: Int?,
        val stops: List<StopUpdate>,
    )

    private data class Snapshot(
        val source: String,
        val headerTimestamp: Long,
        val expiresAt: Long,
        val updates: Map<String, TripUpdate>,
    )

    private val sourceNames = listOf("sydneytrains", "nswtrains", "metro", "ferries", "mff")
    private val snapshots = ConcurrentHashMap<String, Snapshot>()
    private val etags = ConcurrentHashMap<String, String>()

    suspend fun refresh(baseUrl: String) = coroutineScope {
        sourceNames.map { source -> async(Dispatchers.IO) { runCatching { fetch(baseUrl, source) } } }.awaitAll()
    }

    fun hasFreshData(): Boolean = snapshots.values.any { it.expiresAt > System.currentTimeMillis() }

    fun overlay(connections: List<ScheduledConnection>, assignment: (String, String) -> StopAssignment?): List<ScheduledConnection> {
        if (!hasFreshData()) return connections
        val result = connections.toMutableList()
        connections.indices.groupBy { connections[it].tripKey }.values.forEach { indices ->
            val first = connections[indices.first()]
            val snapshot = freshSnapshot(first.source) ?: return@forEach
            val update = snapshot.updates[key(first.tripId, first.serviceDate)] ?: return@forEach
            var carriedDelay = update.delaySeconds?.times(1_000L)
            for (index in indices.sortedBy { connections[it].fromSequence }) {
                val connection = connections[index]
                val from = update.stops.firstOrNull { it.matches(connection.fromStopId, connection.fromSequence) }
                val to = update.stops.firstOrNull { it.matches(connection.toStopId, connection.toSequence) }
                val departure = estimate(from, connection.departure, departure = true, carriedDelay)
                if (from != null && from.relationship != "noData" && departure != null) carriedDelay = departure - connection.departure
                val arrival = estimate(to, connection.arrival, departure = false, carriedDelay)
                if (to != null && to.relationship != "noData" && arrival != null) carriedDelay = arrival - connection.arrival
                val replacementMissingStop = update.status == "replacement" && (from == null || to == null)
                val fromAssignment = from?.assignedStopId?.let { assignment(connection.source, it) }
                val toAssignment = to?.assignedStopId?.let { assignment(connection.source, it) }
                val crossHubAssignment = fromAssignment?.stationId?.let { it != connection.fromStationId } == true ||
                    toAssignment?.stationId?.let { it != connection.toStationId } == true
                result[index] = connection.copy(
                    estimatedDeparture = departure,
                    estimatedArrival = arrival,
                    pickupType = if (from?.relationship == "skipped") 1 else connection.pickupType,
                    dropOffType = if (to?.relationship == "skipped") 1 else connection.dropOffType,
                    fromPlatform = fromAssignment?.platform ?: connection.fromPlatform,
                    toPlatform = toAssignment?.platform ?: connection.toPlatform,
                    cancelled = update.status == "cancelled" || replacementMissingStop || crossHubAssignment || from?.relationship == "skipped" || to?.relationship == "skipped",
                )
            }
        }
        return result
    }

    fun observation(journey: Journey): RealtimeResult<Journey> {
        var observedAt: Long? = null
        var matched = false
        for (identity in journey.legs.mapNotNull(Leg::identity)) {
            val snapshot = freshSnapshot(identity.source) ?: continue
            if (snapshot.updates[key(identity.tripId, identity.serviceDate)] == null) continue
            matched = true
            observedAt = minOf(observedAt ?: Long.MAX_VALUE, snapshot.headerTimestamp)
        }
        return RealtimeResult(journey, observedAt, matched)
    }

    fun overlay(connection: ScheduledConnection, assignment: (String, String) -> StopAssignment?): RealtimeResult<ScheduledConnection> {
        val snapshot = freshSnapshot(connection.source) ?: return RealtimeResult(connection)
        val update = snapshot.updates[key(connection.tripId, connection.serviceDate)] ?: return RealtimeResult(connection)
        val from = update.stops.firstOrNull { it.matches(connection.fromStopId, connection.fromSequence) }
        val to = update.stops.firstOrNull { it.matches(connection.toStopId, connection.toSequence) }
        val defaultDelay = update.delaySeconds?.times(1_000L)
        val departure = if (from?.relationship == "noData") null else from?.departure
            ?: from?.departureDelaySeconds?.let { connection.departure + it * 1_000L }
            ?: defaultDelay?.let { connection.departure + it }
        val arrival = if (to?.relationship == "noData") null else to?.arrival
            ?: to?.arrivalDelaySeconds?.let { connection.arrival + it * 1_000L }
            ?: defaultDelay?.let { connection.arrival + it }
        val replacementMissingStop = update.status == "replacement" && (from == null || to == null)
        val fromAssignment = from?.assignedStopId?.let { assignment(connection.source, it) }
        val toAssignment = to?.assignedStopId?.let { assignment(connection.source, it) }
        val crossHubAssignment = fromAssignment?.stationId?.let { it != connection.fromStationId } == true ||
            toAssignment?.stationId?.let { it != connection.toStationId } == true
        val cancelled = update.status == "cancelled" || replacementMissingStop || crossHubAssignment || from?.relationship == "skipped" || to?.relationship == "skipped"
        val value = connection.copy(
            estimatedDeparture = departure,
            estimatedArrival = arrival,
            pickupType = if (from?.relationship == "skipped") 1 else connection.pickupType,
            dropOffType = if (to?.relationship == "skipped") 1 else connection.dropOffType,
            fromPlatform = fromAssignment?.platform ?: connection.fromPlatform,
            toPlatform = toAssignment?.platform ?: connection.toPlatform,
            cancelled = cancelled,
        )
        return RealtimeResult(value, snapshot.headerTimestamp, true)
    }

    private fun estimate(stop: StopUpdate?, scheduled: Long, departure: Boolean, inheritedDelay: Long?): Long? {
        if (stop?.relationship == "noData") return null
        val exact = if (departure) stop?.departure else stop?.arrival
        val delay = if (departure) stop?.departureDelaySeconds ?: stop?.arrivalDelaySeconds else stop?.arrivalDelaySeconds ?: stop?.departureDelaySeconds
        return exact ?: delay?.let { scheduled + it * 1_000L } ?: inheritedDelay?.let { scheduled + it }
    }

    fun overlay(journey: Journey, assignment: (String, String) -> StopAssignment?): RealtimeResult<Journey> {
        var observedAt: Long? = null
        var matched = false
        val legs = journey.legs.map { leg ->
            val identity = leg.identity ?: return@map leg
            val snapshot = freshSnapshot(identity.source)
                ?: return@map leg.copy(estimatedDeparture = null, estimatedArrival = null, cancelled = false)
            val update = snapshot.updates[key(identity.tripId, identity.serviceDate)]
                ?: return@map leg.copy(estimatedDeparture = null, estimatedArrival = null, cancelled = false)
            matched = true
            observedAt = minOf(observedAt ?: Long.MAX_VALUE, snapshot.headerTimestamp)
            val from = update.stops.firstOrNull { it.matches(identity.fromStopId, identity.fromSequence) }
            val to = update.stops.firstOrNull { it.matches(identity.toStopId, identity.toSequence) }
            val delay = update.delaySeconds?.times(1_000L)
            val replacementMissingStop = update.status == "replacement" && (from == null || to == null)
            val fromAssignment = from?.assignedStopId?.let { assignment(identity.source, it) }
            val toAssignment = to?.assignedStopId?.let { assignment(identity.source, it) }
            val crossHubAssignment = fromAssignment?.stationId?.let { it != leg.from.id } == true ||
                toAssignment?.stationId?.let { it != leg.to.id } == true
            leg.copy(
                estimatedDeparture = if (from?.relationship == "noData") null else from?.departure ?: from?.departureDelaySeconds?.let { leg.departure + it * 1_000L } ?: delay?.let { leg.departure + it },
                estimatedArrival = if (to?.relationship == "noData") null else to?.arrival ?: to?.arrivalDelaySeconds?.let { leg.arrival + it * 1_000L } ?: delay?.let { leg.arrival + it },
                fromPlatform = fromAssignment?.platform ?: leg.fromPlatform,
                toPlatform = toAssignment?.platform ?: leg.toPlatform,
                cancelled = update.status == "cancelled" || replacementMissingStop || crossHubAssignment || from?.relationship == "skipped" || to?.relationship == "skipped",
            )
        }
        return RealtimeResult(Journey(legs), observedAt, matched)
    }

    private fun freshSnapshot(source: String): Snapshot? {
        val snapshot = snapshots[source] ?: return null
        return snapshot.takeIf { it.expiresAt > System.currentTimeMillis() }
    }

    private fun fetch(baseUrl: String, source: String) {
        val url = URL("${baseUrl.trimEnd('/')}/api/v1/realtime/$source")
        require(url.protocol == "https" || url.host in setOf("localhost", "127.0.0.1")) { "Realtime updates require HTTPS" }
        val connection = url.openConnection() as HttpURLConnection
        connection.connectTimeout = 5_000
        connection.readTimeout = 8_000
        connection.setRequestProperty("Accept", "application/json")
        etags[source]?.let { connection.setRequestProperty("If-None-Match", it) }
        try {
            if (connection.responseCode == HttpURLConnection.HTTP_NOT_MODIFIED) return
            if (connection.responseCode != HttpURLConnection.HTTP_OK || connection.getHeaderField("X-Data-Stale").equals("true", true)) return
            val bytes = connection.inputStream.use { readLimited(it, 5_000_000) }
            if (!accept(bytes.toString(Charsets.UTF_8), source)) return
            connection.getHeaderField("ETag")?.let { etags[source] = it }
        } finally {
            connection.disconnect()
        }
    }

    internal fun accept(json: String, expectedSource: String, now: Long = System.currentTimeMillis()): Boolean {
        val snapshot = parseSnapshot(JSONObject(json), expectedSource)
        if (snapshot.expiresAt <= now) return false
        snapshots[expectedSource] = snapshot
        return true
    }

    private fun parseSnapshot(json: JSONObject, expectedSource: String): Snapshot {
        require(json.getInt("schemaVersion") == 1)
        val source = json.getString("source")
        require(source == expectedSource)
        val updates = linkedMapOf<String, TripUpdate>()
        val values = json.getJSONArray("updates")
        for (index in 0 until values.length()) {
            val item = values.getJSONObject(index)
            val tripId = item.getString("tripId")
            val serviceDate = item.getString("serviceDate")
            require(serviceDate.matches(Regex("\\d{8}")))
            val status = item.getString("status")
            require(status in setOf("scheduled", "added", "unscheduled", "cancelled", "replacement"))
            val stops = item.getJSONArray("stopUpdates").let { array ->
                List(array.length()) { stopIndex ->
                    val stop = array.getJSONObject(stopIndex)
                    StopUpdate(
                        stopId = stop.optionalString("stopId"),
                        sequence = stop.optionalInt("stopSequence"),
                        assignedStopId = stop.optionalString("assignedStopId"),
                        arrival = stop.optionalTime("arrivalMs"),
                        departure = stop.optionalTime("departureMs"),
                        arrivalDelaySeconds = stop.optionalInt("arrivalDelaySeconds"),
                        departureDelaySeconds = stop.optionalInt("departureDelaySeconds"),
                        relationship = stop.optString("scheduleRelationship", "scheduled"),
                    )
                }
            }
            val key = key(tripId, serviceDate)
            require(key !in updates)
            updates[key] = TripUpdate(tripId, serviceDate, status, item.optionalInt("delaySeconds"), stops)
        }
        return Snapshot(source, json.requiredTime("headerTimestamp"), json.requiredTime("expiresAt"), updates)
    }

    private fun key(tripId: String, serviceDate: String) = "$tripId\u0000$serviceDate"

    private fun readLimited(input: InputStream, limit: Int): ByteArray {
        val output = ByteArrayOutputStream()
        val buffer = ByteArray(DEFAULT_BUFFER_SIZE)
        while (true) {
            val read = input.read(buffer)
            if (read < 0) break
            require(output.size() + read <= limit)
            output.write(buffer, 0, read)
        }
        return output.toByteArray()
    }

    private fun JSONObject.optionalString(name: String): String? = if (has(name) && !isNull(name)) getString(name).takeIf(String::isNotBlank) else null
    private fun JSONObject.optionalInt(name: String): Int? = if (has(name) && !isNull(name)) getInt(name) else null
    private fun JSONObject.optionalTime(name: String): Long? = if (has(name) && !isNull(name)) parseTime(get(name)) else null
    private fun JSONObject.requiredTime(name: String): Long = parseTime(get(name))
    private fun parseTime(value: Any): Long = when (value) {
        is Number -> value.toLong()
        is String -> Instant.parse(value).toEpochMilli()
        else -> error("invalid timestamp")
    }
}

File: ios/ILoveTrains/Core/OfflineRealtime.swift
import Foundation

struct RealtimeResult<Value> {
    var value: Value
    var observedAt: Millis?
    var matched: Bool
    var matchedLegIndices: Set<Int>

    init(value: Value, observedAt: Millis? = nil, matched: Bool = false, matchedLegIndices: Set<Int> = []) {
        self.value = value
        self.observedAt = observedAt
        self.matched = matched
        self.matchedLegIndices = matchedLegIndices
    }
}

struct StopAssignment: Equatable, Sendable {
    var platform: String?
    var stationId: String
}

struct OfflineRealtime {
    private struct StopUpdate: Decodable, Sendable {
        var stopId: String?
        var stopSequence: Int?
        var assignedStopId: String?
        var arrivalMs: Millis?
        var departureMs: Millis?
        var arrivalDelaySeconds: Int?
        var departureDelaySeconds: Int?
        var scheduleRelationship: String

        enum CodingKeys: String, CodingKey {
            case stopId, stopSequence, assignedStopId, arrivalMs, departureMs
            case arrivalDelaySeconds, departureDelaySeconds, scheduleRelationship
        }

        init(from decoder: Decoder) throws {
            let values = try decoder.container(keyedBy: CodingKeys.self)
            stopId = try values.decodeIfPresent(String.self, forKey: .stopId).flatMap { $0.isEmpty ? nil : $0 }
            stopSequence = try values.decodeIfPresent(Int.self, forKey: .stopSequence)
            assignedStopId = try values.decodeIfPresent(String.self, forKey: .assignedStopId).flatMap { $0.isEmpty ? nil : $0 }
            arrivalMs = try values.decodeIfPresent(FlexibleMillis.self, forKey: .arrivalMs)?.value
            departureMs = try values.decodeIfPresent(FlexibleMillis.self, forKey: .departureMs)?.value
            arrivalDelaySeconds = try values.decodeIfPresent(Int.self, forKey: .arrivalDelaySeconds)
            departureDelaySeconds = try values.decodeIfPresent(Int.self, forKey: .departureDelaySeconds)
            scheduleRelationship = try values.decodeIfPresent(String.self, forKey: .scheduleRelationship) ?? "scheduled"
            guard ["scheduled", "skipped", "noData", "unscheduled"].contains(scheduleRelationship) else {
                throw OfflineCoreError.invalidRealtime
            }
        }

        func matches(id: String, sequence: Int?) -> Bool {
            switch (stopId, stopSequence) {
            case let (.some(stopId), .some(stopSequence)):
                return stopId == id && stopSequence == sequence
            case let (.some(stopId), nil):
                return stopId == id
            case let (nil, .some(stopSequence)):
                return stopSequence == sequence
            case (nil, nil):
                return false
            }
        }
    }

    private struct TripUpdate: Decodable, Sendable {
        var tripId: String
        var serviceDate: String
        var status: String
        var delaySeconds: Int?
        var stopUpdates: [StopUpdate]

        enum CodingKeys: CodingKey { case tripId, serviceDate, status, delaySeconds, stopUpdates }

        init(from decoder: Decoder) throws {
            let values = try decoder.container(keyedBy: CodingKeys.self)
            tripId = try values.decode(String.self, forKey: .tripId)
            serviceDate = try values.decode(String.self, forKey: .serviceDate)
            status = try values.decode(String.self, forKey: .status)
            delaySeconds = try values.decodeIfPresent(Int.self, forKey: .delaySeconds)
            stopUpdates = try values.decode([StopUpdate].self, forKey: .stopUpdates)
            guard serviceDate.count == 8,
                  serviceDate.allSatisfy(\.isNumber),
                  ["scheduled", "added", "unscheduled", "cancelled", "replacement"].contains(status) else {
                throw OfflineCoreError.invalidRealtime
            }
        }
    }

    private struct WireSnapshot: Decodable, Sendable {
        var schemaVersion: Int
        var source: String
        var headerTimestamp: FlexibleMillis
        var expiresAt: FlexibleMillis
        var updates: [TripUpdate]
    }

    private struct Snapshot: Sendable {
        var source: String
        var headerTimestamp: Millis
        var expiresAt: Millis
        var updates: [String: TripUpdate]
    }

    private struct FetchResult: Sendable {
        var source: String
        var data: Data?
        var etag: String?
    }

    private let sourceNames = ["sydneytrains", "nswtrains", "metro", "ferries", "mff"]
    private var snapshots: [String: Snapshot] = [:]
    private var etags: [String: String] = [:]

    mutating func refresh(baseURL: String) async {
        let existingETags = etags
        let sources = sourceNames
        await withTaskGroup(of: FetchResult?.self) { group in
            for source in sources {
                group.addTask {
                    try? await Self.fetch(baseURL: baseURL, source: source, etag: existingETags[source])
                }
            }
            for await result in group {
                guard let result, let data = result.data,
                      (try? accept(data: data, expectedSource: result.source)) == true else { continue }
                if let etag = result.etag { etags[result.source] = etag }
            }
        }
    }

    func hasFreshData(now: Millis = Date().timeIntervalSince1970 * 1_000) -> Bool {
        snapshots.values.contains { $0.expiresAt > now }
    }

    func overlay(
        _ connections: [ScheduledConnection],
        now: Millis = Date().timeIntervalSince1970 * 1_000,
        assignment: (String, String) -> StopAssignment?
    ) -> [ScheduledConnection] {
        guard hasFreshData(now: now) else { return connections }
        var result = connections
        let grouped = Dictionary(grouping: connections.indices) { connections[$0].tripKey }
        for indices in grouped.values {
            guard let firstIndex = indices.first else { continue }
            let first = connections[firstIndex]
            guard let snapshot = freshSnapshot(first.source, now: now),
                  let update = snapshot.updates[Self.key(first.tripId, first.serviceDate)] else { continue }
            var carriedDelay = update.delaySeconds.map { Millis($0) * 1_000 }
            for index in indices.sorted(by: { connections[$0].fromSequence < connections[$1].fromSequence }) {
                let connection = connections[index]
                let from = update.stopUpdates.first { $0.matches(id: connection.fromStopId, sequence: connection.fromSequence) }
                let to = update.stopUpdates.first { $0.matches(id: connection.toStopId, sequence: connection.toSequence) }
                let departure = estimate(from, scheduled: connection.departure, departure: true, inheritedDelay: carriedDelay)
                if let from, from.scheduleRelationship != "noData", let departure {
                    carriedDelay = departure - connection.departure
                }
                let arrival = estimate(to, scheduled: connection.arrival, departure: false, inheritedDelay: carriedDelay)
                if let to, to.scheduleRelationship != "noData", let arrival {
                    carriedDelay = arrival - connection.arrival
                }
                let fromAssignment = from?.assignedStopId.flatMap { assignment(connection.source, $0) }
                let toAssignment = to?.assignedStopId.flatMap { assignment(connection.source, $0) }
                var value = connection
                value.estimatedDeparture = departure
                value.estimatedArrival = arrival
                if from?.scheduleRelationship == "skipped" { value.pickupType = 1 }
                if to?.scheduleRelationship == "skipped" { value.dropOffType = 1 }
                value.fromPlatform = fromAssignment?.platform ?? connection.fromPlatform
                value.toPlatform = toAssignment?.platform ?? connection.toPlatform
                value.cancelled = update.status == "cancelled"
                    || (update.status == "replacement" && (from == nil || to == nil))
                    || fromAssignment.map { $0.stationId != connection.fromStationId } == true
                    || toAssignment.map { $0.stationId != connection.toStationId } == true
                result[index] = value
            }
        }
        return result
    }

    func observation(_ journey: Journey, now: Millis = Date().timeIntervalSince1970 * 1_000) -> RealtimeResult<Journey> {
        var observedAt: Millis?
        var matched = false
        for identity in journey.legs.compactMap(\.identity) {
            guard let snapshot = freshSnapshot(identity.source, now: now),
                  snapshot.updates[Self.key(identity.tripId, identity.serviceDate)] != nil else { continue }
            matched = true
            observedAt = min(observedAt ?? .greatestFiniteMagnitude, snapshot.headerTimestamp)
        }
        return RealtimeResult(value: journey, observedAt: observedAt, matched: matched)
    }

    func overlay(
        _ journey: Journey,
        now: Millis = Date().timeIntervalSince1970 * 1_000,
        assignment: (String, String) -> StopAssignment?
    ) -> RealtimeResult<Journey> {
        var observedAt: Millis?
        var matched = false
        var matchedLegIndices = Set<Int>()
        let legs = journey.legs.enumerated().map { index, leg -> Leg in
            guard let identity = leg.identity,
                  let snapshot = freshSnapshot(identity.source, now: now),
                  let update = snapshot.updates[Self.key(identity.tripId, identity.serviceDate)] else {
                return leg.scheduledOnly()
            }
            matched = true
            matchedLegIndices.insert(index)
            observedAt = min(observedAt ?? .greatestFiniteMagnitude, snapshot.headerTimestamp)
            let from = update.stopUpdates.first { $0.matches(id: identity.fromStopId, sequence: identity.fromSequence) }
            let to = update.stopUpdates.first { $0.matches(id: identity.toStopId, sequence: identity.toSequence) }
            let delay = update.delaySeconds.map { Millis($0) * 1_000 }
            let fromAssignment = from?.assignedStopId.flatMap { assignment(identity.source, $0) }
            let toAssignment = to?.assignedStopId.flatMap { assignment(identity.source, $0) }
            var value = leg
            value.estimatedDeparture = from?.scheduleRelationship == "noData" ? nil
                : from?.departureMs
                    ?? from?.departureDelaySeconds.map { leg.departure + Millis($0) * 1_000 }
                    ?? delay.map { leg.departure + $0 }
            value.estimatedArrival = to?.scheduleRelationship == "noData" ? nil
                : to?.arrivalMs
                    ?? to?.arrivalDelaySeconds.map { leg.arrival + Millis($0) * 1_000 }
                    ?? delay.map { leg.arrival + $0 }
            value.fromPlatform = fromAssignment?.platform ?? leg.fromPlatform
            value.toPlatform = toAssignment?.platform ?? leg.toPlatform
            value.cancelled = update.status == "cancelled"
                || (update.status == "replacement" && (from == nil || to == nil))
                || fromAssignment.map { $0.stationId != leg.from.id } == true
                || toAssignment.map { $0.stationId != leg.to.id } == true
                || from?.scheduleRelationship == "skipped"
                || to?.scheduleRelationship == "skipped"
            return value
        }
        return RealtimeResult(
            value: Journey(legs: legs),
            observedAt: observedAt,
            matched: matched,
            matchedLegIndices: matchedLegIndices
        )
    }

    @discardableResult
    mutating func accept(
        json: String,
        expectedSource: String,
        now: Millis = Date().timeIntervalSince1970 * 1_000
    ) throws -> Bool {
        try accept(data: Data(json.utf8), expectedSource: expectedSource, now: now)
    }

    @discardableResult
    private mutating func accept(
        data: Data,
        expectedSource: String,
        now: Millis = Date().timeIntervalSince1970 * 1_000
    ) throws -> Bool {
        let wire = try JSONDecoder().decode(WireSnapshot.self, from: data)
        guard wire.schemaVersion == 1, wire.source == expectedSource else { throw OfflineCoreError.invalidRealtime }
        var updates: [String: TripUpdate] = [:]
        for update in wire.updates {
            let key = Self.key(update.tripId, update.serviceDate)
            guard updates[key] == nil else { throw OfflineCoreError.invalidRealtime }
            updates[key] = update
        }
        let headerTimestamp = wire.headerTimestamp.value
        let snapshot = Snapshot(
            source: wire.source,
            headerTimestamp: headerTimestamp,
            expiresAt: min(wire.expiresAt.value, headerTimestamp + Self.maximumSnapshotAge),
            updates: updates
        )
        guard snapshot.headerTimestamp <= now + Self.maximumFutureHeaderAge,
              snapshot.expiresAt > now else { return false }
        if let current = snapshots[expectedSource], current.headerTimestamp > snapshot.headerTimestamp { return false }
        snapshots[expectedSource] = snapshot
        return true
    }

    private func freshSnapshot(_ source: String, now: Millis) -> Snapshot? {
        snapshots[source].flatMap { $0.expiresAt > now ? $0 : nil }
    }

    private func estimate(
        _ stop: StopUpdate?,
        scheduled: Millis,
        departure: Bool,
        inheritedDelay: Millis?
    ) -> Millis? {
        if stop?.scheduleRelationship == "noData" { return nil }
        let exact = departure ? stop?.departureMs : stop?.arrivalMs
        let delay = departure
            ? stop?.departureDelaySeconds ?? stop?.arrivalDelaySeconds
            : stop?.arrivalDelaySeconds ?? stop?.departureDelaySeconds
        return exact ?? delay.map { scheduled + Millis($0) * 1_000 } ?? inheritedDelay.map { scheduled + $0 }
    }

    private static func key(_ tripId: String, _ serviceDate: String) -> String {
        "\(tripId)\0\(serviceDate)"
    }

    private static let maximumSnapshotAge: Millis = 90_000
    private static let maximumFutureHeaderAge: Millis = 5 * 60_000

    private static func fetch(baseURL: String, source: String, etag: String?) async throws -> FetchResult {
        guard var components = URLComponents(string: baseURL) else { throw OfflineCoreError.insecureURL }
        let basePath = components.path.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        components.path = (basePath.isEmpty ? "" : "/\(basePath)") + "/api/v1/realtime/\(source)"
        guard let url = components.url, secure(url) else { throw OfflineCoreError.insecureURL }
        var request = URLRequest(url: url, timeoutInterval: 8)
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        if let etag { request.setValue(etag, forHTTPHeaderField: "If-None-Match") }
        let (temporary, response) = try await URLSession.shared.download(for: request)
        defer { try? FileManager.default.removeItem(at: temporary) }
        guard let http = response as? HTTPURLResponse else { throw OfflineCoreError.invalidResponse }
        if http.statusCode == 304 { return FetchResult(source: source, data: nil, etag: nil) }
        let attributes = try FileManager.default.attributesOfItem(atPath: temporary.path)
        guard let size = attributes[.size] as? NSNumber, size.intValue <= 5_000_000 else {
            throw OfflineCoreError.invalidResponse
        }
        let data = try Data(contentsOf: temporary)
        guard http.statusCode == 200,
              http.value(forHTTPHeaderField: "X-Data-Stale")?.lowercased() != "true" else {
            throw OfflineCoreError.invalidResponse
        }
        return FetchResult(source: source, data: data, etag: http.value(forHTTPHeaderField: "ETag"))
    }
}

struct FlexibleMillis: Decodable, Sendable {
    var value: Millis

    init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        if let number = try? container.decode(Double.self) {
            guard validTransitMillis(number) else { throw OfflineCoreError.invalidTimestamp }
            value = number
            return
        }
        let text = try container.decode(String.self)
        let fractional = ISO8601DateFormatter()
        fractional.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        let basic = ISO8601DateFormatter()
        guard let date = fractional.date(from: text) ?? basic.date(from: text) else {
            throw OfflineCoreError.invalidTimestamp
        }
        let timestamp = date.timeIntervalSince1970 * 1_000
        guard validTransitMillis(timestamp) else { throw OfflineCoreError.invalidTimestamp }
        value = timestamp
    }
}

enum OfflineCoreError: Error, Equatable {
    case notInitialized
    case invalidManifest
    case invalidPackage
    case invalidDatabase
    case invalidRealtime
    case invalidTimestamp
    case invalidResponse
    case insecureURL
    case missingBundledResource(String)
    case sqlite(String)
    case updateInProgress
}

private func secure(_ url: URL) -> Bool {
    url.scheme == "https" || (url.scheme == "http" && ["localhost", "127.0.0.1"].contains(url.host))
}
